### PR TITLE
Tests: fix init.lua tests hanging

### PR DESCRIPTION
**Pull request by mjlbach**
Without this workaround, the test suite hangs for me on macOS, linux, and NixOS.

### CODE REVIEW

**Comment by mjlbach** [comment]
@dm1try do you know why this is necessary? These tests seem awfully slow.

**Comment by dm1try** [comment]
@mjlbach I was not able to debug this properly. I remember the test was hanging and waiting for "any" input, so I left the comment.

**Comment by mjlbach** [comment]
Is there any harm in adding the workaround for these tests too?

**Comment by dm1try** [comment]
> Is there any harm in adding the workaround for these tests too?

nope (maybe a "mental" harm only for code reader :))
though, let me re-check why this is needed in the first place.

**Comment by mjlbach** [comment]
What's super weird is this only started giving me issues (very) recently), and doesn't seem to affect CI.

**Comment by dm1try** [comment]
> What's super weird is this only started giving me issues (very) recently), and doesn't seem to affect CI.

yeah, I was not able to reproduce this on my side(osx) on current master

Screen Shot 2021-01-13 at 03 37 26

and I realized why the "workaround" was needed in the first place: the screen on test nvim istance is small enough so the message "Conflicting configs ..." takes multiple lines and shows "Press Enter or type any command.." at the end.

so this not a workaround anymore but an expected step :)

**Comment by mjlbach** [comment]
Now it all makes sense. Thanks :)

**Comment by mjlbach** [comment]
Maybe I should remove the "why is this needed" in the tests haha. I think it's noted somewhere you need CR for small multiline messages

**Comment by dm1try** [comment]
> Maybe I should remove the "why is this needed" in the tests haha. I think it's noted somewhere you need CR for small multiline messages

oops, I've already opened a PR :)

**Comment by mjlbach** [comment]
I'm going to do it in this one anyways since we need to add the additional <CR> not to fail local tests

File: test/functional/core/startup_spec.lua [hunk 1]
```diff
@@ -458,7 +458,7 @@ describe('user config init', function()
 
   it('loads init.lua from XDG config home by default', function()
     clear{ args_rm={'-u' }, env={ XDG_CONFIG_HOME=xconfig }}
-
+    feed('<cr>') -- confirm "Conflicting config ..." message
```
Comment (dm1try) [inline test/functional/core/startup_spec.lua:461]: hmm, in this test, such message should not be exist on clear setup(`init.vim` is not created in this context).

Comment (dm1try) [reply test/functional/core/startup_spec.lua:461]: how did you run the tests? does test `Xhome` directory get removed between test runs?

Comment (mjlbach) [reply test/functional/core/startup_spec.lua:461]: I just ran with make test. Let me see if adding this back causes the test to hang. When I tested before, I had to disable this test in order for the test suite to not hang.

Comment (mjlbach) [reply test/functional/core/startup_spec.lua:461]: If I remove that one, it hangs.

Comment (dm1try) [reply test/functional/core/startup_spec.lua:461]: I was unable to reproduce it, try to debug on your side.
for example, put some "test" expectation after `feed`
```
feed('<cr>') 
matches('check it!', meths.exec('messages', true))
```

Comment (fkm3) [reply test/functional/core/startup_spec.lua:461]: Not sure if helpful, but this test was hanging for me as well. I tried the debugging suggestion and got the following output:

```
[ RUN      ] user config init loads init.lua from XDG config home by default: 1020.91 ms ERR
test/helpers.lua:100: Pattern does not match.
Pattern:
check it!
Actual:
E484: Can't open file $HOME/bin/share/nvim/syntax/syntax.vim

stack traceback:
        test/helpers.lua:100: in function 'matches'
        test/functional/core/startup_spec.lua:463: in function <test/functional/core/startup_spec.lua:459>
```

I was surprised to see it was referencing `$HOME/bin`. I had run `make CMAKE_INSTALL_PREFIX=$HOME/bin/` to build but hadn't run `make install` yet. Running `make install` fixed the hanging test. Unsure if this is working as intended.

Comment (dm1try) [reply test/functional/core/startup_spec.lua:461]: yeah, there is certainly a problem with setup of testing environment(maybe `XDG_CONFIG_HOME` somehow leaked)
I'm going to take a look at this one more time.

Comment (teto) [reply test/functional/core/startup_spec.lua:461]: @dm1try any news ? I am hit by this too. Error is triggered when neovim sees both an init.lua and init.vim but I haven't looked into why it finds both (leftover from previous state ? )

Comment (dm1try) [reply test/functional/core/startup_spec.lua:461]: >  any news ?

@teto I checked the assumption above(about `XDG_CONFIG_HOME`) but didn't find anything suspicious.

> Error is triggered when neovim sees both an init.lua and init.vim but I haven't looked into why it finds both (leftover from previous state ? )

if this is still reproducible , could you re-check if `XDG_CONFIG_HOME` points to "Xhome/Xconfig"
and is this directory removed on each test run(maybe there is some os-specific behavior and [this setup](https://github.com/neovim/neovim/blob/2dd383c943ea31abf30ec1e39fbeb5f5971131aa/test/functional/core/startup_spec.lua#L442-L457) does not work as expected)
 
also, maybe the tests are run in parallel somehow?)

Comment (mjlbach) [reply test/functional/core/startup_spec.lua:461]: Having some trouble reproducing the hangups

**Comment by mjlbach** [comment]
@dm1try I did some additional digging. The issue seems to be the runtime path getting polluted with entries from $HOME.

See [with my standard neovim configuration](https://paste.sr.ht/~mjlbach/9bb2677e4b27b221ddd23bfbc086109e85865cc9)
And [where I deleted ~/.local/share/nvim](https://paste.sr.ht/~mjlbach/4523055157c42e3b1428330a0a95e1e5579e5b03#nvim.log%20(no%20user%20data%20dir)%20test%20freeze%20on%20user%20config%20init%20loads%20init.lua%20from%20XDG%20config%20home%20by%20default-L27)

Note the runtime/packpath. Note, just deleting local packpath is not sufficient to prevent the test from hanging, I believe the hang is due to [this](https://paste.sr.ht/~mjlbach/4523055157c42e3b1428330a0a95e1e5579e5b03#nvim.log%20(no%20user%20data%20dir)%20test%20freeze%20on%20user%20config%20init%20loads%20init.lua%20from%20XDG%20config%20home%20by%20default-L5) error about missing syntax.vim in runtime path.

The following modification fixes the test with feeding <cr>, but these aren't the correct directories.

```lua
  it('loads init.lua from XDG config home by default', function()
    clear{args_rm={'-u' }, env={ XDG_CONFIG_HOME=xconfig ; XDG_DATA_HOME=xconfig; VIMRUNTIME=xconfig}}

    eq(1, eval('g:lua_rc'))
    eq(init_lua_path, eval('$MYVIMRC'))
  end)
```

**Comment by mjlbach** [comment]
Sidenote, I'm also mildly concerned about the complaints about the ["Failed to start server: address in use"](https://paste.sr.ht/~mjlbach/4523055157c42e3b1428330a0a95e1e5579e5b03#nvim.log%20(no%20user%20data%20dir)%20test%20freeze%20on%20user%20config%20init%20loads%20init.lua%20from%20XDG%20config%20home%20by%20default-L3) warning.

File: test/functional/core/startup_spec.lua [hunk 2]
```diff
@@ -441,6 +441,7 @@ describe('user config init', function()
 
   before_each(function()
     rmdir(xhome)
+    rmdir(xconfig)
```
Comment (mjlbach) [inline test/functional/core/startup_spec.lua:444]: Removing xconfig isn't necessary to fix the tests, but this seems like something we should do for consistency? @teto

File: test/functional/core/startup_spec.lua [hunk 3]
```diff
@@ -472,7 +473,7 @@ describe('user config init', function()
     end)
 
     it('loads custom lua config and does not set $MYVIMRC', function()
-      clear{ args={'-u', custom_lua_path }, env={ XDG_CONFIG_HOME=xconfig }}
+      clear{ args={'-u', custom_lua_path }, env={ XDG_CONFIG_HOME=xconfig ; XDG_DATA_HOME=xconfig; VIMRUNTIME=xconfig}}
```
Comment (mjlbach) [inline test/functional/core/startup_spec.lua:476]: What's weird is XDG_DATA_HOME should be set on line 12 in cmake `set(ENV{XDG_DATA_HOME} ${BUILD_DIR}/Xtest_xdg/share)`

Comment (mjlbach) [reply test/functional/core/startup_spec.lua:476]: I guess this points at `INFO  2021-02-18T11:07:10.488 1012134 emsg_multiline:681: E484: Can't open file Xhome/Xconfig/syntax/syntax.vim`

Comment (mjlbach) [reply test/functional/core/startup_spec.lua:476]: Here's the full RTP path *without* manually overriding XDG_DATA_HOME:
```
INFO  2021-02-18T11:09:34.293 1014444 set_init_2:754: startup runtimepath/packpath value: Xhome/Xconfig/nvim,/etc/xdg/nvim,Xhome/Xconfig/nvim/site,/nix/store/mzbylj1zprn18m7h2nsf6vmfafkvz54r-cmake-3.19.3/share/nvim/site,/nix/store/yn0mz3d30ryx5ids1lf3c4jxgi7r2rci-gettext-0.21/share/nvim/site,/nix/store/6c2s51yjkb2nab010b8dhdldh28dhnvj-pkg-config-wrapper-0.29.2/share/nvim/site,/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/share/nvim/site,/nix/store/khx9absslbabgfwbzxig5258nh138k1p-include-what-you-use-0.14/share/nvim/site,/nix/store/whmd7hgfnfy27ab8d9m7z0d05v0h0yai-fish-3.1.2/share/nvim/site,/nix/store/fbkni3i1sjfxnfld8f6dmrri1f0q5y6i-gnugrep-3.6/share/nvim/site,/nix/store/r19q2xb4qw4l9v13hi1zgjrw5sd5dg0x-gnused-4.8/share/nvim/site,/nix/store/f89071fz20w31wjf116h5l3m5j8qkwld-groff-1.22.4/share/nvim/site,/nix/store/8i2wxn1gbjvaq0l68yhz4dsfpdnkai59-man-db-2.9.3/share/nvim/site,/nix/store/c03izgzw69sz495r2ck3admhq01p0q05-nodejs-14.15.5/share/nvim/site,/nix/store/jm7kyn357z32hz5nr1pgkpgyfyiai4b0-python-2.7.18-env/share/nvim/site,/nix/store/mcxhdbhz0wqi3icx6lvc33iaggd02v6v-patchelf-0.12/share/nvim/site,Xhome/Xconfig,/home/michael/Repositories/neovim_development/neovim/build/lib/nvim,/nix/store/mcxhdbhz0wqi3icx6lvc33iaggd02v6v-patchelf-0.12/share/nvim/site/after,/nix/store/jm7kyn357z32hz5nr1pgkpgyfyiai4b0-python-2.7.18-env/share/nvim/site/after,/nix/store/c03izgzw69sz495r2ck3admhq01p0q05-nodejs-14.15.5/share/nvim/site/after,/nix/store/8i2wxn1gbjvaq0l68yhz4dsfpdnkai59-man-db-2.9.3/share/nvim/site/after,/nix/store/f89071fz20w31wjf116h5l3m5j8qkwld-groff-1.22.4/share/nvim/site/after,/nix/store/r19q2xb4qw4l9v13hi1zgjrw5sd5dg0x-gnused-4.8/share/nvim/site/after,/nix/store/fbkni3i1sjfxnfld8f6dmrri1f0q5y6i-gnugrep-3.6/share/nvim/site/after,/nix/store/whmd7hgfnfy27ab8d9m7z0d05v0h0yai-fish-3.1.2/share/nvim/site/after,/nix/store/khx9absslbabgfwbzxig5258nh138k1p-include-what-you-use-0.14/share/nvim/site/after,/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/share/nvim/site/after,/nix/store/6c2s51yjkb2nab010b8dhdldh28dhnvj-pkg-config-wrapper-0.29.2/share/nvim/site/after,/nix/store/yn0mz3d30ryx5ids1lf3c4jxgi7r2rci-gettext-0.21/share/nvim/site/after,/nix/store/mzbylj1zprn18m7h2nsf6vmfafkvz54r-cmake-3.19.3/share/nvim/site/after,Xhome/Xconfig/nvim/site/after,/etc/xdg/nvim/after,Xhome/Xconfig/nvim/after
INFO  2021-02-18T11:09:34.307 1014444 emsg_multiline:681: E484: Can't open file Xhome/Xconfig/syntax/syntax.vim
```

Comment (mjlbach) [reply test/functional/core/startup_spec.lua:476]: With: 
```
INFO  2021-02-18T11:11:48.061 1016497 set_init_2:754: startup runtimepath/packpath value: Xhome/Xconfig/nvim,/etc/xdg/nvim,Xhome/Xconfig/nvim/site,/nix/store/mzbylj1zprn18m7h2nsf6vm
fafkvz54r-cmake-3.19.3/share/nvim/site,/nix/store/yn0mz3d30ryx5ids1lf3c4jxgi7r2rci-gettext-0.21/share/nvim/site,/nix/store/6c2s51yjkb2nab010b8dhdldh28dhnvj-pkg-config-wrapper-0.29.2
/share/nvim/site,/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/share/nvim/site,/nix/store/khx9absslbabgfwbzxig5258nh138k1p-include-what-you-use-0.14/share/nvim/site,/nix
/store/whmd7hgfnfy27ab8d9m7z0d05v0h0yai-fish-3.1.2/share/nvim/site,/nix/store/fbkni3i1sjfxnfld8f6dmrri1f0q5y6i-gnugrep-3.6/share/nvim/site,/nix/store/r19q2xb4qw4l9v13hi1zgjrw5sd5dg0
x-gnused-4.8/share/nvim/site,/nix/store/f89071fz20w31wjf116h5l3m5j8qkwld-groff-1.22.4/share/nvim/site,/nix/store/8i2wxn1gbjvaq0l68yhz4dsfpdnkai59-man-db-2.9.3/share/nvim/site,/nix/s
tore/c03izgzw69sz495r2ck3admhq01p0q05-nodejs-14.15.5/share/nvim/site,/nix/store/jm7kyn357z32hz5nr1pgkpgyfyiai4b0-python-2.7.18-env/share/nvim/site,/nix/store/mcxhdbhz0wqi3icx6lvc33i
aggd02v6v-patchelf-0.12/share/nvim/site,Xhome/Xconfig,/home/michael/Repositories/neovim_development/neovim/build/lib/nvim,/nix/store/mcxhdbhz0wqi3icx6lvc33iaggd02v6v-patchelf-0.12/s
hare/nvim/site/after,/nix/store/jm7kyn357z32hz5nr1pgkpgyfyiai4b0-python-2.7.18-env/share/nvim/site/after,/nix/store/c03izgzw69sz495r2ck3admhq01p0q05-nodejs-14.15.5/share/nvim/site/a
fter,/nix/store/8i2wxn1gbjvaq0l68yhz4dsfpdnkai59-man-db-2.9.3/share/nvim/site/after,/nix/store/f89071fz20w31wjf116h5l3m5j8qkwld-groff-1.22.4/share/nvim/site/after,/nix/store/r19q2xb
4qw4l9v13hi1zgjrw5sd5dg0x-gnused-4.8/share/nvim/site/after,/nix/store/fbkni3i1sjfxnfld8f6dmrri1f0q5y6i-gnugrep-3.6/share/nvim/site/after,/nix/store/whmd7hgfnfy27ab8d9m7z0d05v0h0yai-
fish-3.1.2/share/nvim/site/after,/nix/store/khx9absslbabgfwbzxig5258nh138k1p-include-what-you-use-0.14/share/nvim/site/after,/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.
7/share/nvim/site/after,/nix/store/6c2s51yjkb2nab010b8dhdldh28dhnvj-pkg-config-wrapper-0.29.2/share/nvim/site/after,/nix/store/yn0mz3d30ryx5ids1lf3c4jxgi7r2rci-gettext-0.21/share/nv
im/site/after,/nix/store/mzbylj1zprn18m7h2nsf6vmfafkvz54r-cmake-3.19.3/share/nvim/site/after,Xhome/Xconfig/nvim/site/after,/etc/xdg/nvim/after,Xhome/Xconfig/nvim/after
INFO  2021-02-18T11:11:48.071 1016497 emsg_multiline:681: E484: Can't open file Xhome/Xconfig/syntax/syntax.vim
```

**Comment by mjlbach** [comment]
Ok, weird, the syntax.vim error still pops up if I set XDG_DATA_HOME, but the test doesn't hang...

**Comment by mjlbach** [comment]
Ok, the issue is actually solely due to XDG_DATA_HOME loading user configs

**Comment by mjlbach** [comment]
This actually seems to be an issue with not running installPhase. Using the nix flake

```bash
cmakeConfigurephase
buildPhase
TEST_FILE="test/functional/core/startup_spec.lua" make functionaltest # hangs
installPhase
TEST_FILE="test/functional/core/startup_spec.lua" make functionaltest # fails
```

**Comment by dm1try** [comment]
> @dm1try I did some additional digging. The issue seems to be the runtime path getting polluted with entries from $HOME, and missing syntax.vim (as also noted by @fkm3)

good investigation
not sure about the `nix` issue described above, but it looks like [`runtimepath`](https://github.com/neovim/neovim/blob/master/runtime/doc/options.txt#L4772-L4789) should be properly stubbed in test environment for excluding side-effects provided by sharing those directories with other environments/instances.

File: test/functional/core/startup_spec.lua (hunk 2 above)
Comment (dm1try) [reply test/functional/core/startup_spec.lua:444]: why? it's a nested directory.

**Comment by mjlbach** [comment]
Closing as we know what the issue is (seems like a nix thing)

**Comment by teto** [comment]
I dont think it's a nix issue, the tests are looking into the install folder (I beleive CMAKE_INSTALL_PREFIX) but one should be able to run the tests without installation. We may need to better control the runtimepath in the tests.

**Comment by mjlbach** [comment]
Sorry, that was a bit dismissive on my part. I should have said, that the issue is exposed by nix, and we should probably address what you and I had discussed on IRC, but that isn't really about fixing the test, but fixing CMAKE_INSTALL_PREFIX so we can run tests without initialization

**Pull request closed without merging**